core/domain/learner_group_services.py

```diff
@@ -0,0 +1,633 @@
+            }
+            matching_story_syllabus_item_dicts.append(syllabus_story_dict)
+
+    return matching_story_syllabus_item_dicts
+
+
+def add_student_to_learner_group(
+    group_id: str,
+    user_id: str,
+    progress_sharing_permission: bool
+) -> None:
+    """Adds the given student to the given learner group.
+
+    Args:
+        group_id: str. The id of the learner group.
+        user_id: str. The id of the student.
+        progress_sharing_permission: bool. The progress sharing permission of
+            the learner group. True if progress sharing is allowed, False
+            otherwise.
+
+    Raises:
+        Exception. Student was not invited to join the learner group.
+    """
+    learner_group_model = learner_group_models.LearnerGroupModel.get(
+        group_id, strict=True
+    )
+
+    if user_id not in learner_group_model.invited_student_user_ids:
+        raise Exception('Student was not invited to join the learner group.')
+
+    learner_group_model.invited_student_user_ids.remove(user_id)
+    learner_group_model.student_user_ids.append(user_id)
+
+    details_of_learner_group = {
+        'group_id': group_id,
+        'progress_sharing_is_turned_on': progress_sharing_permission
+    }
+
+    learner_grps_user_model = user_models.LearnerGroupsUserModel.get(
+        user_id, strict=True
+    )
+
+    learner_grps_user_model.invited_to_learner_groups_ids.remove(group_id)
+    learner_grps_user_model.learner_groups_user_details.append(
+        details_of_learner_group)
+
+    learner_grps_user_model.update_timestamps()
+    learner_grps_user_model.put()
+
+    learner_group_model.update_timestamps()
+    learner_group_model.put()
+
+
+def invite_students_to_learner_group(
+    group_id: str,
+    invited_student_ids: List[str]
+) -> None:
+    """Invites the given students to the given learner group.
+
+    Args:
+        group_id: str. The id of the learner group.
+        invited_student_ids: list(str). The ids of the students to invite.
+    """
+    learner_groups_user_models = (
+        user_models.LearnerGroupsUserModel.get_multi(invited_student_ids))
+
+    for index, student_id in enumerate(invited_student_ids):
+        learner_groups_user_model = learner_groups_user_models[index]
+        if learner_groups_user_model:
+            learner_groups_user_model.invited_to_learner_groups_ids.append(
+                group_id)
+        else:
+            learner_groups_user_model = user_models.LearnerGroupsUserModel(
+                id=student_id,
+                invited_to_learner_groups_ids=[group_id],
+                learner_groups_user_details=[]
+            )
+
+        learner_groups_user_models[index] = learner_groups_user_model
+
+    user_models.LearnerGroupsUserModel.update_timestamps_multi( # type: ignore[type-var]
+        learner_groups_user_models)
+    user_models.LearnerGroupsUserModel.put_multi(learner_groups_user_models) # type: ignore[type-var]
```

Comment:
Why are these ignores needed?

Reply:
They were causing mypy errors as learner_groups_user_models could have none type objects, have modified this a bit and it's not required now. Thanks!